src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java

```diff
@@ -43,36 +48,68 @@ public static NumberRange<Double> searchMinMaxScaleDenominator(Set<StyleInfo> st
             throws IOException {
         // searches the maximum and minimum denominator in the style's rules that are contained in
         // the style set.
-        double minScaleDenominator = Double.POSITIVE_INFINITY;
-        double maxScaleDenominator = Double.NEGATIVE_INFINITY;
+        MinMaxDenominator minMaxDenominator = new MinMaxDenominator();
+        minMaxDenominator.setMinScaleDenominator(Double.POSITIVE_INFINITY);
+        minMaxDenominator.setMaxScaleDenominator(Double.NEGATIVE_INFINITY);
 
         for (StyleInfo styleInfo : styles) {
-            Style style = styleInfo.getStyle();
-            for (FeatureTypeStyle fts : style.featureTypeStyles()) {
-
-                for (Rule rule : fts.rules()) {
-
-                    if (rule.getMinScaleDenominator() < minScaleDenominator) {
-                        minScaleDenominator = rule.getMinScaleDenominator();
+            if (styleInfo.getSLD() != null && styleInfo.getSLD().getStyledLayers() != null) {
+                for (StyledLayer styledLayer : styleInfo.getSLD().getStyledLayers()) {
```

Comment:
The path towards ``styleInfo.getSLD().getStyledLayers()``  is traversed multiple times. Maybe:

```java
Optional<List<StyledLayer>> styledLayers = Optional.ofNullable(styleInfo.getSLD()).map(sld -> sld.getStyledLayers());
if (styledLayers.isPresent()) {
   for (StyledLayer styledLayer : styledLayers.get()) {
   ...
   }
} else if (styleInfo.getStyle() != null) {
   ...
}
```

Reply:
Done